src/block_monitor.rs

```diff
@@ -43,14 +43,41 @@ impl BlockMonitorDisplayOptions {
 #[derive(Clone)]
 pub struct BlockMonitor<C: subxt::Config> {
 	listener_request_tx: mpsc::Sender<(HashOf<C>, TxFoundListenerTrigger<HashOf<C>>)>,
-	_p: PhantomData<C>,
 }
 
 #[async_trait]
 impl<C: subxt::Config> TransactionMonitor<HashOf<C>> for BlockMonitor<C> {
-	async fn wait(&self, tx_hash: HashOf<C>) -> HashOf<C> {
-		let listener = self.register_listener(tx_hash).await;
-		listener.await.unwrap()
+	async fn wait(
+		&self,
+		tx_hash: HashOf<C>,
+		tx_mortality: &Option<u64>,
+	) -> Result<HashOf<C>, Error> {
+		let callback = async {
+			let listener = self.register_listener(tx_hash).await;
+			listener
+				.await
+				.map_err(|err| Error::Other(format!("error while listening for block hash: {err}")))
+		};
+
+		// Wait under timeout only for mortal txs
+		if let Some(mortality) = tx_mortality {
+			tokio::time::timeout(
+				// Consider that 10 blocks are worth a minute
+				Duration::from_secs((mortality / 10 + 1) * 60),
```

Comment:
I think it would be more reasonable to add mortality property to the listener, and just count the timeout using actual block numbers that passed (plus some safety buffer on top of that if we don't know exact starting block number). Assuming 6s/block is not correct (e.g. yap parachain withe elastic scaling).

Reply:
Agreed. I'll make this more robust

Reply:
I accounted for how many finalized blocks to wait for, in the case for mortal txs, but I am checking the last finalized block on chain regularly at 6s interval.